Input/Input.js

```diff
@@ -226,50 +251,72 @@ const InputBase = kind({
 	},
 
 	handlers: {
-		onChange: (ev, {onChange}) => {
-			if (onChange) {
-				onChange({value: ev.target.value});
-			}
-		}
+		onChange: handle(
+			adaptEvent(
+				ev => ({
+					type: 'onBeforeChange',
+					value: ev.target.value
+				}),
+				forwardWithPrevent('onBeforeChange')
+			),
+			forwardCustom('onChange', ev => ({
+				stopPropagation: () => ev.stopPropagation(),
+				value: ev.target.value
+			}))
+		)
 	},
 
 	computed: {
-		// 'aria-label': ({placeholder, type, value}) => {
-		// 	const title = (value == null || value === '') ? placeholder : '';
-		// 	return calcAriaLabel(title, type, value);
-		// },
-		className: ({focused, invalid, styler}) => styler.append({focused, invalid}),
+		'aria-label': ({placeholder, type, value}) => {
+			const title = (value == null || value === '') ? placeholder : '';
+			return calcAriaLabel(title, type, value);
+		},
+		className: ({focused, invalid, size, styler}) => styler.append({focused, invalid}, size),
+		dir: ({value, placeholder}) => isRtlText(value || placeholder) ? 'rtl' : 'ltr',
+		invalidTooltip: ({css, invalid, invalidMessage = $L('Please enter a valid value.')}) => {
+			if (invalid && invalidMessage) {
+				return (
+					<Tooltip css={css} relative>
+						{invalidMessage}
+					</Tooltip>
+				);
+			}
+		},
 		// ensure we have a value so the internal <input> is always controlled
 		value: ({value}) => typeof value === 'number' ? value : (value || '')
 	},
 
-	render: ({css, disabled, iconAfter, iconBefore, onChange, placeholder, size, type, value, ...rest}) => {
+	render: ({css, dir, disabled, iconAfter, iconBefore, iconSize, invalidTooltip, onChange, placeholder, size, type, value, ...rest}) => {
 		const inputProps = extractInputProps(rest);
 		delete rest.dismissOnEnter;
 		delete rest.focused;
 		delete rest.invalid;
 		delete rest.invalidMessage;
+		delete rest.onBeforeChange;
 		delete rest.rtl;
 
 		return (
 			<div {...rest} aria-disabled={disabled} disabled={disabled}>
-				<InputDecoratorIcon position="before" size={size}>
+				<InputDecoratorIcon position="before" size={iconSize}>
 					{iconBefore}
 				</InputDecoratorIcon>
 				<input
 					{...inputProps}
 					aria-disabled={disabled}
 					className={css.input}
+					dir={dir}
 					disabled={disabled}
 					onChange={onChange}
 					placeholder={placeholder}
+					size={size}
```

Comment:
This is API breaking!

Reply:
You're right. I reverted the change and now the `size` prop changes both input and icon sizes.
I also fixed the icon padding. Thanks!